dapr/proto/components/v1/pubsub.proto

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+package dapr.proto.components.v1;
+
+import "dapr/proto/components/v1/common.proto";
+
+option go_package = "github.com/dapr/dapr/pkg/proto/components/v1;components";
+
+// PubSub service provides a gRPC interface for pubsub components.
+service PubSub {
+  // Initializes the pubsub component with the given metadata.
+  rpc Init(PubSubInitRequest) returns (PubSubInitResponse) {}
+
+  // Returns a list of implemented pubsub features.
+  rpc Features(FeaturesRequest) returns (FeaturesResponse) {}
+
+  // Publish publishes a new message for the given topic.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+
+  // Establishes a stream with the server, which sends messages down to the
+  // client. The client streams acknowledgements back to the server. The server
+  // will close the stream and return the status on any error. In case of closed
+  // connection, the client should re-establish the stream.
+  // The message MUST contain a `subscription` on it that should be used for the
+  // entire streaming pull.
+  rpc PullMessages(stream PullMessagesRequest)
+      returns (stream PullMessageResponse) {}
```

Comment:
I would say that PullMessages is good enough because we don't even have any mention to "Subscribe", should I rename subscription to `topic` that might be clear?

```proto
// Used for acknowledge a message.
message PullMessagesRequest {
  // Required. The topic for which to initialize the new stream. This
  // must be provided in the first request on the stream, and must not be set in
  // subsequent requests from client to server.
  Topic topic = 1;
  // The unique message ID.
  string ack_message_id = 2;
  // Optional, should not be fulfilled when the message was successfully
  // handled.
  AckMessageError ack_error = 3;
}

message Topic {
  // The topic name desired to be subscribed
  string name = 1;
  // Metadata related subscribe request.
  map<string, string> metadata = 2;
}
```

how that's sound @tmacam ?

Reply:
I changed to `topic`. I found it even better because I think now it's not ambiguous with "subscription" attribute, the associated semantics of message's order became clearer, don't you think?

Reply:
What I was missing is the Subscribe part of the Publish Subscribe component in the definition of this GRPC service. That's all.

Reply:
I agree with you, seems subscribe method is missing, let me explain the difference with the subscription semantics

when you invoke subscribe on code you have to provide a callback/handler function that will be called when new messages are available, the subscribe function itself should not block.

which is semantically different from PullMessages that’s expected to block until stream ends

Thats why I changed to topic to avoid associating this call as not blocking

So instead of publish/subscribe we have

push[toRopic]
pull[fromTopic]

may I rename publish to push?
So we can eliminate the subscription semantics?

Reply:
I like this push/pull idea. It leaks a lot of the implementation details into de naming of these methods but I am OK with that. As you pointed out, it makes it clear that there is a blocking semantic at play with those two calls.